File: 1046.last-stone-weight.java
// @leet start
class Solution {
    public int lastStoneWeight(int[] stones) {
        PriorityQueue<Integer> queue = new PriorityQueue<>(Collections.reverseOrder());

        for (int stone : stones) {
            queue.add(stone);
        }
        
        while (queue.size() > 1) {
            int first = queue.poll();
            int second = queue.poll();
            int diff = first - second;
            if (diff > 0) {
                queue.add(diff);
            }
        }
        return queue.size() == 0 ? 0 : queue.poll();
    }
}
// @leet end

File: 1046.last-stone-weight.cpp
// @leet start
class Solution {
public:
    int lastStoneWeight(vector<int>& stones) {
        std::make_heap(stones.begin(), stones.end());

        while(stones.size() > 1) {
            std::pop_heap(stones.begin(), stones.end());
            int a = stones.back();
            stones.pop_back();
            std::pop_heap(stones.begin(), stones.end());
            int b = stones.back();
            stones.pop_back();
            if (a == b) {
                continue;
            }
            stones.push_back(a - b)
            std::push_heap(stones.begin(), stones.end());
        }

        return stones.size() == 0 ? 0 : stones[0];
    }
};
// @leet end
